autode/opt/optimisers/base.py

```diff
@@ -890,78 +847,360 @@ def print_geometries(self, filename: Optional[str] = None) -> None:
+        """
+        Retrieve the stored optimiser parameters from the trajectory
+        file
+
+        Returns:
+            (dict): Dictionary of optimiser parameters
+        """
+        assert self._filename is not None
+        with ZipFile(self._filename, "r") as file:
+            names = file.namelist()
 
-          |
-        E |    -----   /          <-- Does contain a rise in the energy
-          |         \/
-          |_________________
-               Iteration
+        if "opt_params" not in names:
+            return dict()
+
+        with ZipFile(self._filename, "r") as file:
+            params = file.read("opt_params").decode("utf-8")
+
+        all_params = NumericStringDict(params)
+        opt_params = {}
+        for key in ["maxiter", "gtol", "etol"]:
+            res = all_params.get(key, None)
+            if res is not None:
+                opt_params[key] = res
+
+        return opt_params
+
+    def add(self, coords: "OptCoordinates") -> None:
+        """
+        Add a new set of coordinates to this trajectory
+
+        Args:
+            coords (OptCoordinates): The set of coordinates to be added
+        """
+        if coords is None:
+            return None
+        elif not isinstance(coords, OptCoordinates):
+            raise ValueError("item added must be OptCoordinates")
+
+        # check if we need to push last coords to disk or can skip
+        if len(self._memory) < 2 or self._len < self._n_stored + 2:
+            self._memory.append(coords)
+            self._len += 1
+            return None
+
+        if self._filename is None:
+            raise RuntimeError(
+                "Must initialise the OptimiserHistory object"
+                "before adding more than two coordinates"
+            )
+
+        n_stored = self._n_stored
+        with ZipFile(self._filename, "a") as file:
+            with file.open(f"coords_{n_stored}", "w") as fh:
+                fh.write(self._coords_to_bytes(self.penultimate))
+        self._memory.append(coords)
+        self._len += 1
+        return None
+
+    def flush(self):
+        """
+        Put the coordinates that are in memory on disk
+        """
+        # prevent duplication of coords on disk
+        n_stored = self._n_stored
+        n_to_flush = self._len - self._n_stored
+        assert n_to_flush <= 2
+        if n_to_flush == 0:
+            return None
+        elif n_to_flush == 1:
+            idxs = [n_stored]
+            coords_list = [self.final]
+        else:
+            idxs = [n_stored, n_stored + 1]
+            coords_list = [self.penultimate, self.final]
+
+        with ZipFile(self._filename, "a") as file:
+            for idx, coords in zip(idxs, coords_list):
+                with file.open(f"coords_{idx}", "w") as fh:
+                    fh.write(self._coords_to_bytes(coords))
+        return None
+
+    @staticmethod
+    def _coords_to_bytes(coords: OptCoordinates) -> bytes:
```

Comment:
this is lots of effort. why not pickle?

Reply:
@t-young31 Pickling is said to be unsafe in the sense that it can run arbitrary code while unpickling. Given the feature of reading back an old saved trajectory from disk, it seems better to serialise manually. What do you think?